### ISSUE TITLE
In test stage, the infer audio signals are different even thought I archived them from the same checkpoint.

### ISSUE THREAD

**Issue by njusq** [facebookresearch/demucs#523]
## ❓ Questions
I extract the test progress in utils/solver.py and use it to test the same checkpoint for several times. 
```
308     def get_epoch_from_ckpt(self, current_ckpt):
309         pattern = r"\d+"
310         matches = re.findall(pattern, current_ckpt)
311         if matches:
312             current_epoch = int(matches[-1])
313             return current_epoch
314         raise NotImplementedError
315
316     def test(self, res_file, current_ckpt):
317         if self.history:
318             logger.info("Replaying metrics from previous run")
319         for epoch, metrics in enumerate(self.history):
320             formatted = self._format_train(metrics['train'])
321             logger.info(
322                 bold(f'Train Summary | Epoch {epoch + 1} | {_summary(formatted)}'))
323             formatted = self._format_train(metrics['valid'])
324             logger.info(
325                 bold(f'Valid Summary | Epoch {epoch + 1} | {_summary(formatted)}'))
326             if 'test' in metrics:
327                 formatted = self._format_test(metrics['test'])
328                 if formatted:
329                     logger.info(bold(f"Test Summary | Epoch {epoch + 1} | {_summary(formatted)}"))
330         # Evaluate on the testset
331         logger.info('-' * 70)
332         logger.info('Evaluating on the test set...')
333         # We switch to the best known model for testing
334         if self.args['test']['best']:
335             state = self.best_state
336         else:
337             state = states.copy_state(self.model.state_dict())
338         compute_sdr = self.args['test']['sdr']
339         current_epoch = self.get_epoch_from_ckpt(current_ckpt)
340         with states.swap_state(self.model, state):
341             with torch.no_grad():
342                 metrics['test'] = evaluate(self, compute_sdr=compute_sdr)
343         formatted = self._format_test(metrics['test'])
344         print(metrics['test'])
345         testline = f"Test Summary | Epoch {current_epoch} | {_summary(formatted)}"
346         logger.info(bold(testline))
347         # update test results
348         if distrib.rank == 0:
349             with open(res_file, "a+") as f:
350                 f.write(testline + "\n")
```

But I found that the results are different. e.g. For Epoch 359, the results show that
```
Test Summary | Epoch 359 | Sdr=6.146 | Nsdr=6.401 | Sdr_drums=7.592 | Nsdr_drums=7.645 | Sdr_bass=6.947 | Nsdr_bass=6.981 | Sdr_other=3.983 | Nsdr_other=4.271 | Sdr_vocals=6.061 | Nsdr_vocals=6.706
Test Summary | Epoch 359 | Sdr=6.142 | Nsdr=6.394 | Sdr_drums=7.593 | Nsdr_drums=7.648 | Sdr_bass=6.919 | Nsdr_bass=6.956 | Sdr_other=3.979 | Nsdr_other=4.266 | Sdr_vocals=6.075 | Nsdr_vocals=6.706
Test Summary | Epoch 359 | Sdr=6.191 | Nsdr=6.456 | Sdr_drums=7.542 | Nsdr_drums=7.631 | Sdr_bass=6.933 | Nsdr_bass=6.975 | Sdr_other=4.128 | Nsdr_other=4.444 | Sdr_vocals=6.161 | Nsdr_vocals=6.771
```
I also set conf/config.yaml to save inference audio in test stage, and found that the inference audio signal are different each time.
```
test:
  save: True
``` 

Could you tell me what introduced the differences? How could I do to avoid the difference to make the result reproductive?

**Comment by adefossez** [comment]
See https://pytorch.org/docs/stable/notes/randomness.html for details on perfect reproducibility. this will come at a cost though!

**Comment by njusq** [comment]
Glad to see your response!
I figure out my bug, which occurred in the segment parameter when inference.
I used the configuration file conf/config.yaml the same as which in your repository and the setting is
`use_train_segment: true`. 
The forward function in htdemucs.py impacts the segment parameter.
```
        if self.use_train_segment:
            if self.training:
                self.segment = Fraction(mix.shape[-1], self.samplerate)
            else:
                training_length = int(self.segment * self.samplerate)
                if mix.shape[-1] < training_length:
                    length_pre_pad = mix.shape[-1]
                    mix = F.pad(mix, (0, training_length - length_pre_pad))
```
In your training and testing process, we will train the model before testing and then the segment will be set as `self.segment = Fraction(mix.shape[-1], self.samplerate)`.
However, in my process I test the model directly without training step before testing.The self.segment will use the value in configuration instead of the value `self.segment = Fraction(mix.shape[-1], self.samplerate)` fixed in training step. After I fixed the segment value in conf/config.yaml, the test result is the same as I got in train function from solver.py.
